lib/streamlit/runtime/secrets.py

```diff
@@ -45,33 +44,47 @@ def _missing_key_error_message(key: str) -> str:
     )
 
 
-class AttrDict(UserDict):  # type: ignore[type-arg]
+class AttrDict(Mapping[str, Any]):
     """
     We use AttrDict to wrap up dictionary values from secrets
     to provide dot access to nested secrets
     """
 
+    def __init__(self, value):
+        self.__dict__.update(value)
+
     @staticmethod
     def _maybe_wrap_in_attr_dict(value) -> Any:
         if not isinstance(value, dict):
             return value
         else:
-            return AttrDict(**value)
+            return AttrDict(value)
 
-    def __getattr__(self, attr_name: str) -> Any:
-        try:
-            value = super().__getitem__(attr_name)
-            return self._maybe_wrap_in_attr_dict(value)
-        except KeyError:
-            raise AttributeError(_missing_attr_error_message(attr_name))
+    def __len__(self) -> int:
+        return len(self.__dict__)
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self.__dict__)
 
     def __getitem__(self, key: str) -> Any:
         try:
-            value = super().__getitem__(key)
+            value = self.__dict__[key]
             return self._maybe_wrap_in_attr_dict(value)
         except KeyError:
             raise KeyError(_missing_key_error_message(key))
 
+    def __getattr__(self, attr_name: str) -> Any:
+        try:
+            return self[attr_name]
+        except KeyError:
+            raise AttributeError(_missing_attr_error_message(attr_name))
+
+    def __setitem__(self, key, value):
```

Comment:
nit: I don't think we need to overwrite `__setitem__` and `__setattr__` since it's not required by mapping and it will return the same error anyways. Or did you do it here to adapt the error message?

Reply:
Correct, I did that to keep exceptions similar to high-level Secrets object, to not show internal machinery (expose `AttrDict`) to the final user.

Reply:
ahh, ok that makes sense. But maybe one aspect to think about: if we add those methods, typing checks will think that they actually exist. So there might be a slight benefit of not adding those, to allow type/linters to directly show it as an error.

Reply:
hm, my checks show that with or without `__setattr__` or `__setitem__` neither PyCharm nor mypy does complain about an attempt to assign to attribute or item. Only pyright complains about `__setitem__` missing during item assignment. 

But without having methods above, user could write something like this

```python
x = st.secrets.servers.alpha
x.some_ip = "7.7.7.7"

st.write(x.some_ip)
```
Which will not work in case `__setattrs__` explicitly raise an exception. 

Anyway, I explicitly added `NoReturn` return type to the methods https://peps.python.org/pep-0484/#the-noreturn-type to explicitly indicate that they should never return successfully. Let's hope some linters in the future will highlight warnings for this.

Reply:
Alright, sounds good :)